ogr/ogrsf_frmts/xodr/ogrxodrdatasource.cpp

```diff
@@ -0,0 +1,186 @@
+ * Purpose:  Implementation of OGRXODRDataSource.
+ * Author:   Michael Scholz, German Aerospace Center (DLR)
+ *           Gülsen Bardak, German Aerospace Center (DLR)        
+ *
+ ******************************************************************************
+ * Copyright 2024 German Aerospace Center (DLR), Institute of Transportation Systems
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "ogr_xodr.h"
+using namespace odr;
+using namespace pugi;
+using namespace std;
+
+CPL_CVSID("$Id$")
+
+OGRXODRDataSource::OGRXODRDataSource()
+{
+    layers = nullptr;
+    nLayers = 0;
+}
+
+OGRXODRDataSource::~OGRXODRDataSource()
+{
+    for (int i = 0; i < nLayers; i++)
+    {
+        delete layers[i];
+        layers[i] = nullptr;
+    }
+    CPLFree(layers);
+}
+
+int OGRXODRDataSource::Open(const char *fileName, char **openOptions,
+                            int bUpdate)
+{
+    VSILFILE *file = nullptr;
+
+    bool updatable = CPL_TO_BOOL(bUpdate);
+    if (updatable)
+    {
+        CPLError(CE_Failure, CPLE_OpenFailed,
+                 "Write or update access not supported by XODR driver.");
+        return CE_Failure;
+    }
+    else
+    {
+        file = VSIFOpenExL(fileName, "r", true);
+    }
+
+    if (file == nullptr)
+    {
+        //TODO is this ever called on an opening error? An incorrect file name or path is caught earlier already.
+        CPLError(CE_Failure, CPLE_OpenFailed,
+                 "Failed to load OpenDRIVE file %s.", fileName);
+        return CE_Failure;
+    }
+
+    const char *openOptionValue =
+        CSLFetchNameValueDef(openOptions, "EPS", "1.0");
+    eps = CPLAtof(openOptionValue);
+    openOptionValue = CSLFetchNameValueDef(openOptions, "DISSOLVE_TIN", "NO");
+    dissolveTIN = CPLTestBool(openOptionValue);
+
+    odr::OpenDriveMap xodr(fileName, false);
```

Comment:
can't this fail?

Reply:
See my [other answer](https://github.com/OSGeo/gdal/pull/9504#issuecomment-2113442286). For now, I will create a basic check if an "OpenDRIVE" node has successfully been parsed during creation of the `odr::OpenDriveMap`. This would fail for non-OpenDRIVE input files.

If the input just contains random structural XML errors, but generally is OpenDRIVE, all content up to the first error will be parsed and the creation of the `odr::OpenDriveMap` will *not* fail, unfortunately.